pkg/controller/controller_instance.go

```diff
@@ -377,6 +379,16 @@ func (c *controller) reconcileServiceInstanceAdd(instance *v1beta1.ServiceInstan
 		// recordStartOfServiceInstanceOperation has updated the instance, so we need to continue in the next iteration
 		return nil
 	}
```

Comment:
This reminded me of something else. 

I'm unclear about what return err vs nil means in the context of this outermost function.

Should the `if instance.Status.CurrentOperation == "" {` be a guard clause, with the `recordStart` taking place fully after the entire of the outer `if` body?

My read is, if we're not doing an operation, start one, but if we are doing an operation, go back to the beginning of the loop.

Is this to avoid updating the existing instance after we've already updated it, because it's been requeued by making a change? Shouldn't we proceed onward and do the rest of the instance processing? Is there something at the beginning of t he function before this that needs to be rechecked?

Reply:
> I'm unclear about what return err vs nil means in the context of this outermost function.

Returning error mostly tells the controller to perform exponential backoff (rate limiter) and not marking item as done.

> Is this to avoid updating the existing instance after we've already updated it, because it's been requeued by making a change?

Yes.

> Shouldn't we proceed onward and do the rest of the instance processing?

No, because there are issues with cache caused by this like #1639
We should only update instance once during a single iteration.

Reply:
@MHBauer In case you (and other) wonder why do we update status in the middle of the process at all. The normal Kube controller logic is to never read the status - it's only used to report to the user. So it's perfectly fine if we do some multi-step work for a minute and then fail to update the status - the operation will be retried and everything's gonna be _eventually_ fine.

We, on the other hand, have to read the status and drive our controller logic based on the last written status. Hence we can't just update the status in the end - we have to commit every stage of the process - effectively we have a "status driven workflow".

The reason for that is that OSB is not Kube friendly - it doesn't have GET endpoints (so we can't get the state of instance/binding), and it's hard to tell which exactly operation is currently in progress (there is just `last_operation` without details of operation itself). I see this as a hole in OSB REST API, and hope that we will eventually fix it in `v3`.